### PR TITLE
fix: honour ancestor BOM/parent dependency management in fix action

### LINKED ISSUE #154: dependencies fix action: hardcodes resolved versions instead of using property refs or honouring ancestor BOM management
## Problem

When running `pilot:dependencies -Dpilot.action=fix` on a multi-module project whose root POM already manages dependency versions via `<dependencyManagement>` (either directly or through properties), the fix action promotes used transitive dependencies correctly but writes **hardcoded literal versions** rather than:

1. **Omitting the `<version>`** when the dep is already managed by an ancestor BOM/parent POM, or
2. **Using the property expression** (e.g. `${resolverVersion}`) when one exists in the inheritance chain.

## Reproduction

Run against Apache Maven master (`4.1.0-SNAPSHOT` @ `4309183ac9`):

```bash
cd /path/to/apache/maven
mvn eu.maveniverse.maven.plugins:pilot-plugin:0.3.0:dependencies -Dpilot.action=fix
```

The root `pom.xml` defines:
```xml
<properties>
  <resolverVersion>2.0.22</resolverVersion>
  <woodstoxVersion>7.2.2</woodstoxVersion>
</properties>
<dependencyManagement>
  <dependencies>
    <dependency>
      <groupId>org.apache.maven.resolver</groupId>
      <artifactId>maven-resolver-named-locks</artifactId>
      <version>${resolverVersion}</version>
    </dependency>
    <!-- ... all other resolver artifacts, woodstox, plexus-sec-dispatcher, etc. -->
  </dependencies>
</dependencyManagement>
```

Pilot produces child module `<dependencyManagement>` blocks like:
```xml
<!-- compat/maven-toolchain-builder/pom.xml — WRONG output -->
<dependencyManagement>
  <dependencies>
    <dependency>
      <groupId>org.apache.maven.resolver</groupId>
      <artifactId>maven-resolver-named-locks</artifactId>
      <version>2.0.22</version>   <!-- should be absent: root BOM already manages this -->
    </dependency>
    <dependency>
      <groupId>org.codehaus.plexus</groupId>
      <artifactId>plexus-sec-dispatcher</artifactId>
      <version>4.2.0</version>   <!-- should be absent: root BOM already manages this -->
    </dependency>
    <dependency>
      <groupId>com.fasterxml.woodstox</groupId>
      <artifactId>woodstox-core</artifactId>
      <version>7.2.2</version>   <!-- should be absent or ${woodstoxVersion} -->
    </dependency>
  </dependencies>
</dependencyManagement>
```

## Expected behaviour

Before adding a `<version>` to a promoted dependency, pilot should walk the **effective dependency management** inherited from parent POMs:

- If the dep is already managed in an ancestor `<dependencyManagement>`, emit it **without `<version>`** (the child's `<dependencyManagement>` entry is unnecessary entirely — ideally pilot skips creating it, or at minimum omits the version tag).
- If the dep is managed via a property expression in an ancestor, preserve that expression rather than resolving it to a literal.

## Affected artifacts (Maven example)

| Artifact | Root BOM entry | Pilot wrote | Expected |
|---|---|---|---|
| `org.apache.maven.resolver:*` | `${resolverVersion}` = 2.0.22 | `2.0.22` hardcoded | absent (BOM covers it) |
| `com.fasterxml.woodstox:woodstox-core` | `${woodstoxVersion}` = 7.2.2 | `7.2.2` hardcoded | absent (BOM covers it) |
| `org.codehaus.plexus:plexus-sec-dispatcher` | literal in root BOM | `4.2.0` hardcoded | absent (BOM covers it) |
| `org.opentest4j:opentest4j` | in root BOM | literal | absent |
| `org.junit.platform:junit-platform-commons` | in root BOM | literal | absent |
| `org.apache.maven:*` (reactor) | reactor-managed | `4.1.0-SNAPSHOT` | also redundant if root BOM covers it |

## Secondary issue: unnecessary child `<dependencyManagement>` blocks

When all of a child module's new `<dependencyManagement>` entries are already covered by the root BOM, the entire `<dependencyManagement>` section pilot creates is superfluous. Pilot should detect this and not emit it at all.

## Note on correctness of the dep set

The **which** deps to add/remove looks directionally correct — the bytecode analysis seems sound. The issue is solely in the **version strategy** for the promoted deps.

### PR DESCRIPTION
**Pull request by gnodet**
Fixes #154. Stacked on #157.

## Problem

When `pilot:dependencies -Dpilot.action=fix` promotes a used transitive dependency to `<dependencies>`, it always hardcodes the **resolved literal version** — even when an ancestor POM's `<dependencyManagement>` already covers that dependency (possibly via a property expression such as `${resolverVersion}`).

This produces incorrect child-module POMs like:

```xml
<dependency>
  <groupId>org.apache.maven.resolver</groupId>
  <artifactId>maven-resolver-named-locks</artifactId>
  <version>2.0.22</version>  <!-- wrong: ancestor BOM already manages this -->
</dependency>
```

## Fix

Before adding a `<version>` element, the fix action now checks whether the dependency is already version-managed by an **ancestor BOM or parent POM** (effective inherited `dependencyManagement` minus what the current module itself declares).

- **Ancestor-managed dep** → emitted without `<version>`. Maven inherits the correct version (and honours any property expression defined in the ancestor).
- **Not ancestor-managed** → previous behaviour preserved; resolved literal version is written.

Leverages the new `addAligned(Coordinates.of(g, a, null), opts)` API introduced in domtrip 1.6.1 (#157): passing a `null` version emits the dependency without a `<version>` element.

## Implementation

| File | Change |
|---|---|
| `DependenciesReporter` | New `fix()` overload accepting `Set<String> ancestorManagedGAs`; uses `addAligned(Coordinates.of(g, a, null))` for ancestor-managed deps. Existing 5-arg overload delegates with empty set (backward-compatible). |
| `DependenciesMojo` | New `buildAncestorManagedGAs()` helper; threads the set through `executeNonInteractive()` to `fix()`. |
| `AnalyzeDependenciesMojo` | Same propagation applied (was using the old 5-arg overload). |
| `DependenciesReporterTest` | 4 new tests: ancestor-managed → no version; ancestor-managed + non-compile scope; non-ancestor-managed → version written; no pre-existing `<dependencies>` section (deps == null branch). |

## Summary by CodeRabbit

- **Bug Fixes**
  - Dependency fixes now recognize versions managed by ancestor POMs or imported BOMs.
  - Transitive dependencies omit redundant `<version>` entries when their versions are managed elsewhere.
  - Existing dependency versions remain unchanged when no ancestor management applies.
  - Non-compile scopes are preserved during dependency fixes.
  - Fixes now also work when the project does not yet contain a `<dependencies>` section.
  - Classified dependencies are correctly recognized when managed by ancestor configuration.

### CODE REVIEW

File: pilot-core/src/main/java/eu/maveniverse/maven/pilot/DependenciesReporter.java [hunk 1]
```diff
@@ -84,12 +86,31 @@ public static String formatCheckFailure(
...
+     *
+     * <p>When adding used transitive dependencies, honours ancestor dependency management:</p>
+     * <ul>
+     *   <li>If the dependency is already managed by an ancestor BOM/parent POM
+     *       ({@code ancestorManagedGAs} contains its GA key), it is added <em>without</em> a
+     *       {@code <version>} element — the inherited management already pins the version.</li>
+     *   <li>Otherwise the resolved version from {@code gaToVersion} is used.  If the resolved
+     *       version equals a value that was originally expressed as a property in an ancestor
+     *       ({@code gaToVersionExpression} is non-null and contains the GA), the property
+     *       expression (e.g. {@code ${resolverVersion}}) is written instead of the literal.</li>
```
[the remark this answers is not in the document]
Comment (gnodet) [reply pilot-core/src/main/java/eu/maveniverse/maven/pilot/DependenciesReporter.java]: Fixed in 6626451. Removed the `gaToVersionExpression` bullet from the Javadoc — the method only writes the literal value from `gaToVersion`.

File: pilot-core/src/test/java/eu/maveniverse/maven/pilot/DependenciesReporterTest.java [hunk 2]
```diff
@@ -239,4 +240,105 @@ void fixBothRemovesAndAdds(@TempDir Path tempDir) throws Exception {
...
+                List.of(),
+                List.of(transitive),
+                Map.of("org.managed:ancestor-lib", "3.0"),
+                ancestorManaged,
+                logs::add);
+
+        String result = Files.readString(pomPath);
+        assertThat(result).contains("ancestor-lib");
+        // Version must NOT appear: ancestor BOM already manages it
+        assertThat(result).doesNotContain("3.0");
```
[the remark this answers is not in the document]
Comment (gnodet) [reply pilot-core/src/test/java/eu/maveniverse/maven/pilot/DependenciesReporterTest.java]: Fixed in 6626451. Assertion tightened to `doesNotContain("<version>X.Y</version>")` (version-specific). Also added `fixAncestorManagedNoDependenciesSection` test covering the `deps == null` branch (POM with no pre-existing `<dependencies>` section).

File: pilot-core/src/main/java/eu/maveniverse/maven/pilot/DependenciesReporter.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (gnodet) [reply pilot-core/src/main/java/eu/maveniverse/maven/pilot/DependenciesReporter.java]: Fixed in 6626451. Removed the stale second `<li>` describing `gaToVersionExpression`/property-expression behaviour that isn't implemented.

File: pilot-plugin/src/main/java/eu/maveniverse/maven/pilot/mvn3/DependenciesMojo.java [hunk 3]
```diff
@@ -183,7 +188,47 @@ private void executeForProject(MavenProject proj) throws Exception {
...
+        Set<String> ancestorManagedGAs = new HashSet<>();
+        if (proj.getDependencyManagement() != null
+                && proj.getDependencyManagement().getDependencies() != null) {
+            for (Dependency dep : proj.getDependencyManagement().getDependencies()) {
+                String classifier = dep.getClassifier();
+                String ga = (classifier != null && !classifier.isEmpty())
+                        ? dep.getGroupId() + ":" + dep.getArtifactId() + ":" + classifier
+                        : dep.getGroupId() + ":" + dep.getArtifactId();
+                if (!ownManagedGAs.contains(ga)) {
+                    ancestorManagedGAs.add(ga);
```
[the remark this answers is not in the document]
Comment (gnodet) [reply pilot-plugin/src/main/java/eu/maveniverse/maven/pilot/mvn3/DependenciesMojo.java]: Acknowledged in 6626451. Added a Javadoc note to `buildAncestorManagedGAs` explicitly documenting this design choice: dependencies supplied by a BOM imported in the current POM are intentionally classified as ancestor-managed. Version-less is correct as long as the BOM import remains present — that is the standard Maven pattern for BOM consumers.

File: pilot-plugin/src/main/java/eu/maveniverse/maven/pilot/mvn3/DependenciesMojo.java [hunk 4]
```diff
@@ -183,7 +189,49 @@ private void executeForProject(MavenProject proj) throws Exception {
...
+     *
+     * <p>When a used-transitive dependency is already managed by an ancestor, the fix action
+     * should add it to {@code <dependencies>} without a {@code <version>} element.</p>
+     *
+     * <p>Note: dependencies provided by a BOM that is <em>imported</em> in this module's own
+     * {@code <dependencyManagement>} (via {@code <scope>import</scope>}) are classified as
+     * ancestor-managed. This is intentional — those dependencies are version-pinned by the BOM
+     * import, so omitting {@code <version>} is correct as long as the import remains present.</p>
+     */
+    static Set<String> buildAncestorManagedGAs(MavenProject proj) {
```
[the remark this answers is not in the document]
Comment (gnodet) [reply pilot-plugin/src/main/java/eu/maveniverse/maven/pilot/mvn3/DependenciesMojo.java:208]: Fixed in 4e1deb4. Added 5 unit tests covering all three key scenarios: own DM entry (excluded), parent-inherited entry (included), null InputLocation (conservatively included as ancestor), plus classifier handling and empty-DM edge case.

File: pilot-plugin/src/main/java/eu/maveniverse/maven/pilot/mvn3/DependenciesMojo.java [hunk 5]
```diff
@@ -183,7 +189,49 @@ private void executeForProject(MavenProject proj) throws Exception {
...
+        Set<String> ancestorManagedGAs = new HashSet<>();
+        if (proj.getModel().getDependencyManagement() == null
+                || proj.getModel().getDependencyManagement().getDependencies() == null) {
+            return ancestorManagedGAs;
+        }
+        for (Dependency dep : proj.getModel().getDependencyManagement().getDependencies()) {
+            InputLocation loc = dep.getLocation("");
+            String sourcePath =
+                    (loc != null && loc.getSource() != null) ? loc.getSource().getLocation() : null;
+            if (sourcePath == null || !sourcePath.equals(ownPomPath)) {
```
[the remark this answers is not in the document]
Comment (gnodet) [reply pilot-plugin/src/main/java/eu/maveniverse/maven/pilot/mvn3/DependenciesMojo.java:231]: Fixed in 56a9129. Normalize `sourcePath` via `Path.of(rawSourcePath).normalize().toString()` before comparing with `ownPomPath`. URL-style locations (containing `://`) are preserved unchanged — they can never match a local absolute path. Regression test `buildAncestorManagedGAs_nonNormalizedOwnPathExcluded` added covering the `.`-segment case.
